File: backend/app/routers/stocks.py
from typing import Literal

from fastapi import APIRouter, Depends, Query

from app.auth import get_current_user
from app.dependencies import get_stock_service
from app.schemas.auth import UserPublic
from app.schemas.stocks import StockInfo, StockSearchItem, TimeSeriesResponse
from app.services import StockService

router = APIRouter(prefix="/stocks", tags=["stocks"])


@router.get("/search", response_model=list[StockSearchItem])
async def search_stocks(
    q: str = Query(min_length=1),
    current_user: UserPublic = Depends(get_current_user),
    service: StockService = Depends(get_stock_service),
) -> list[StockSearchItem]:
    _ = current_user
    return await service.search(q)


@router.get("/{symbol}", response_model=StockInfo)
async def get_stock(
    symbol: str,
    current_user: UserPublic = Depends(get_current_user),
    service: StockService = Depends(get_stock_service),
) -> StockInfo:
    _ = current_user
    return await service.get_stock(symbol)


@router.get("/{symbol}/timeseries", response_model=TimeSeriesResponse)
async def get_timeseries(
    symbol: str,
    mode: Literal["realtime", "historical"] = Query(...),
    interval: Literal["1min", "5min", "15min"] = Query(...),
    start_date: str | None = Query(default=None),
    end_date: str | None = Query(default=None),
    current_user: UserPublic = Depends(get_current_user),
    service: StockService = Depends(get_stock_service),
) -> TimeSeriesResponse:
    _ = current_user
    return await service.get_timeseries(
        symbol,
        mode=mode,
        interval=interval,
        start_date=start_date,
        end_date=end_date,
    )

File: frontend/src/api/stocks.ts
import { apiRequest } from './client'
import type {
  ChartMode,
  Interval,
  StockInfo,
  StockSearchItem,
  TimeSeriesResponse,
} from '../types/api'

export function searchStocks(token: string, q: string): Promise<StockSearchItem[]> {
  const params = new URLSearchParams({ q })
  return apiRequest<StockSearchItem[]>(`/stocks/search?${params}`, { token })
}

export function getStock(token: string, symbol: string): Promise<StockInfo> {
  return apiRequest<StockInfo>(`/stocks/${encodeURIComponent(symbol)}`, { token })
}

export function getTimeseries(
  token: string,
  symbol: string,
  params: {
    mode: ChartMode
    interval: Interval
    start_date?: string
    end_date?: string
  },
): Promise<TimeSeriesResponse> {
  const query = new URLSearchParams({
    mode: params.mode,
    interval: params.interval,
  })
  if (params.start_date) query.set('start_date', params.start_date)
  if (params.end_date) query.set('end_date', params.end_date)

  return apiRequest<TimeSeriesResponse>(
    `/stocks/${encodeURIComponent(symbol)}/timeseries?${query}`,
    { token },
  )
}
